### PR TITLE
Cadet API v2

### PR DESCRIPTION
**Pull request by bnjmnt4n**
Long-running branch for API v2 changes.

### CODE REVIEW

**Comment by qreoct** [comment]
Note: tests don't pass because there are some changes in the [assessments PR](https://github.com/source-academy/cadet/pull/729) not merged here yet. 

We are still awaiting the work there to be finalized.

File: lib/cadet_web/admin_controllers/admin_assessments_controller.ex [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    end
+  end
+
+  def update(conn, params = %{"assessmentid" => assessment_id}) when is_ecto_id(assessment_id) do
+    open_at = params |> Map.get("openAt")
+    close_at = params |> Map.get("closeAt")
+    is_published = params |> Map.get("isPublished")
+
+    updated_assessment =
+      if is_published == nil do
```
Comment (angelsl) [inline lib/cadet_web/admin_controllers/admin_assessments_controller.ex:58]: Not sure if these were in the original code, but please change these to `is_nil(is_published)`

Comment (angelsl) [inline lib/cadet_web/admin_controllers/admin_assessments_controller.ex:58]: Not sure if these were in the original code, but please change these to `is_nil(is_published)`

File: lib/cadet_web/admin_controllers/admin_assessments_controller.ex [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    else
+      {:error, {status, message}} ->
+        conn
+        |> put_status(status)
+        |> text(message)
+    end
+  end
+
+  defp check_dates(open_at, close_at, assessment) do
+    if open_at == nil and close_at == nil do
```
Comment (angelsl) [inline lib/cadet_web/admin_controllers/admin_assessments_controller.ex:76]: `is_nil`

File: test/cadet_web/admin_controllers/admin_assessments_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    end
+  end
+
+  describe "DELETE /:assessment_id, staff only" do
+    @tag authenticate: :staff
+    test "successful", %{conn: conn} do
+      assessment = insert(:assessment)
+      conn = delete(conn, build_url(assessment.id))
+      assert response(conn, 200) == "OK"
+      assert Repo.get(Assessment, assessment.id) == nil
```
Comment (angelsl) [inline test/cadet_web/admin_controllers/admin_assessments_controller_test.exs:141]: `assert is_nil`

File: test/cadet_web/admin_controllers/admin_assessments_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+  end
+
+  describe "POST /:assessment_id, staff only, publish" do
+    @tag authenticate: :staff
+    test "successful toggle from published to unpublished", %{conn: conn} do
+      assessment = insert(:assessment, is_published: true)
+      conn = post(conn, build_url(assessment.id), %{isPublished: false})
+      expected = Repo.get(Assessment, assessment.id).is_published
+      assert response(conn, 200) == "OK"
+      assert expected == false
```
Comment (angelsl) [inline test/cadet_web/admin_controllers/admin_assessments_controller_test.exs:169]: `refute`

File: test/cadet_web/admin_controllers/admin_assessments_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    end
+
+    @tag authenticate: :staff
+    test "successful toggle from unpublished to published", %{conn: conn} do
+      assessment = insert(:assessment, is_published: false)
+      conn = post(conn, build_url(assessment.id), %{isPublished: true})
+      expected = Repo.get(Assessment, assessment.id).is_published
+      assert response(conn, 200) == "OK"
+      assert expected == true
```
Comment (angelsl) [inline test/cadet_web/admin_controllers/admin_assessments_controller_test.exs:178]: Drop the `== true`

**Review by angelsl — COMMENTED** [review]
Quite a big PR, so I've not been able to read everything in detail, but I think it's okay since most of the changes are just moving code around.

Just some minor tweaks.

**Review by angelsl — APPROVED** [review]

**Pull request merged**